File: backend/app/models/enums.py
"""Domain enumerations shared by ORM models and Pydantic schemas.

These mirror the value sets named in the design document's Data Models section
and the requirements (e.g. Role admin/invigilator/student, alert severity
info/warning/danger). Keeping them in one place guarantees the ORM and the
schema mirrors stay in agreement.
"""

from __future__ import annotations

from enum import StrEnum


class Role(StrEnum):
    """The three RBAC roles recognized by the platform (Requirement 2.3)."""

    ADMIN = "admin"
    INVIGILATOR = "invigilator"
    STUDENT = "student"


class ExamStatus(StrEnum):
    """Exam lifecycle states (design Data Models: Exam.status)."""

    DRAFT = "draft"
    PROVISIONING = "provisioning"
    LIVE = "live"
    COMPLETED = "completed"


class QuestionType(StrEnum):
    """Supported question types (design Data Models: Question.type)."""

    MCQ = "mcq"
    SHORT = "short"
    NUMERICAL = "numerical"


class AuditStatus(StrEnum):
    """Auditor verdict states for a generated paper."""

    PENDING = "pending"
    APPROVED = "approved"
    FLAGGED = "flagged"


class SessionStatus(StrEnum):
    """Exam session lifecycle states (design Data Models: ExamSession.status)."""

    NOT_STARTED = "not_started"
    ACTIVE = "active"
    SUBMITTED = "submitted"
    TERMINATED = "terminated"


class SessionEventKind(StrEnum):
    """Telemetry event kinds (design Data Models: SessionEvent.kind)."""

    TAB_BLUR = "tab_blur"
    TAB_FOCUS = "tab_focus"
    PASTE = "paste"
    COPY = "copy"
    ANSWER_CHANGE = "answer_change"
    QUESTION_VIEW = "question_view"
    HEARTBEAT = "heartbeat"


class SourceAgent(StrEnum):
    """Agents that can source an anomaly (design Data Models: Anomaly.source_agent)."""

    GUARDIAN = "guardian"
    SENTINEL = "sentinel"


class AnomalyCategory(StrEnum):
    """Anomaly categories (design Data Models: Anomaly.category)."""

    FACE_ABSENT = "face_absent"
    MULTIPLE_FACES = "multiple_faces"
    GAZE_AWAY = "gaze_away"
    TAB_SWITCH = "tab_switch"
    PASTE = "paste"
    TIMING = "timing"
    ANSWER_SIMILARITY = "answer_similarity"


class AlertSeverity(StrEnum):
    """Alert severity enumeration (Requirement 9.1, 14.8)."""

    INFO = "info"
    WARNING = "warning"
    DANGER = "danger"

File: frontend/src/types/enums.ts
// Domain enumerations mirroring backend `app/models/enums.py`.
// Kept as string-literal unions (the wire format is the enum's string value).

/** The three RBAC roles recognized by the platform (backend Role). */
export type Role = "admin" | "invigilator" | "student";

/** Exam lifecycle states (backend ExamStatus). */
export type ExamStatus = "draft" | "provisioning" | "live" | "completed";

/** Supported question types (backend QuestionType). */
export type QuestionType = "mcq" | "short" | "numerical";

/** Auditor verdict states for a generated paper (backend AuditStatus). */
export type AuditStatus = "pending" | "approved" | "flagged";

/** Exam session lifecycle states (backend SessionStatus). */
export type SessionStatus =
  | "not_started"
  | "active"
  | "submitted"
  | "terminated";

/** Telemetry event kinds (backend SessionEventKind). */
export type SessionEventKind =
  | "tab_blur"
  | "tab_focus"
  | "paste"
  | "copy"
  | "answer_change"
  | "question_view"
  | "heartbeat";

/** Agents that can source an anomaly (backend SourceAgent). */
export type SourceAgent = "guardian" | "sentinel";

/** Anomaly categories (backend AnomalyCategory). */
export type AnomalyCategory =
  | "face_absent"
  | "multiple_faces"
  | "gaze_away"
  | "tab_switch"
  | "paste"
  | "timing"
  | "answer_similarity";

/** Alert severity enumeration (backend AlertSeverity). */
export type AlertSeverity = "info" | "warning" | "danger";
